web/packages/shared/components/UnifiedResources/UnifiedResources.tsx

```diff
@@ -103,58 +102,58 @@ export type UnifiedResourcesPinning =
       kind: 'hidden';
     };
 
-interface UnifiedResourcesProps<T> {
-  params: ResourceFilter;
+interface UnifiedResourcesProps {
+  params: UnifiedResourcesQueryParams;
+  resourcesFetchAttempt: Attempt;
+  fetchResources: (options?: { force?: boolean }) => Promise<void>;
+  resources: SharedUnifiedResource[];
   //TODO(gzdunek): the pin button should be moved to some other place
   //according to the new designs
   Header(pinAllButton: React.ReactElement): React.ReactElement;
-  EmptySearchResults: React.ReactElement;
+  /** Rendered when resource list and search query are empty. */
+  NoResources: React.ReactElement;
   /**
    * If pinning is supported, the functions to get and update pinned resources
    * can be passed here.
    */
   pinning: UnifiedResourcesPinning;
   availableKinds: SharedUnifiedResource['resource']['kind'][];
-  /**
-   * Fetches the resources.
-   * The type of particular resource is not specified; the fetched resource
-   * should be mapped to `SharedUnifiedResource` via the `mapToResource` prop.
-   * */
-  fetchFunc(
-    params: UrlResourcesParams,
-    signal: AbortSignal
-  ): Promise<ResourcesResponse<T>>;
-  /** Maps fetched data to `SharedUnifiedResource`. */
-  mapToResource(response: T): SharedUnifiedResource;
-  setParams(params: ResourceFilter): void;
+  setParams(params: UnifiedResourcesQueryParams): void;
   onLabelClick(label: ResourceLabel): void;
   updateUnifiedResourcesPreferences(
     preferences: UnifiedResourcePreferences
   ): void;
 }
 
-export function UnifiedResources<T>(props: UnifiedResourcesProps<T>) {
+export function useUnifiedResourcesFetch<T>(props: {
+  fetchFunc(
+    paginationParams: { limit: number; startKey: string },
+    signal: AbortSignal
+  ): Promise<ResourcesResponse<T>>;
+}) {
+  return useKeyBasedPagination({
+    fetchFunc: props.fetchFunc,
+    initialFetchSize: INITIAL_FETCH_SIZE,
+    fetchMoreSize: FETCH_MORE_SIZE,
+  });
+}
```

Comment:
I'd move this definition somewhere else so that it's not between `UnifiedResourcesProps` and `UnifiedResources`. Or alternatively just inline `UnifiedResources` props.

Reply:
Moved below the component definition.